ksql-engine/src/main/java/io/confluent/ksql/util/StatementUtil.java

```diff
@@ -0,0 +1,141 @@
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.AbstractStreamCreateStatement;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.IntegerLiteral;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+
+public final class StatementUtil {
+
+  private StatementUtil() {
+  }
+
+  public static <T extends Statement> PreparedStatement<T> withInferredSchema(
+      final T statement,
+      final String statementText,
+      final SchemaRegistryClient schemaRegistryClient) {
+    return withInferredSchema(PreparedStatement.of(statementText, statement), schemaRegistryClient);
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends Statement> PreparedStatement<T> withInferredSchema(
+      final PreparedStatement<T> statement,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    if (!(statement.getStatement() instanceof AbstractStreamCreateStatement)) {
+      return statement;
+    }
+
+    final AbstractStreamCreateStatement statementWithSchema = AvroUtil.checkAndSetAvroSchema(
+        (AbstractStreamCreateStatement) statement.getStatement(), schemaRegistryClient);
+
+    if (statementWithSchema == statement.getStatement()) {
+      return statement;
+    }
+
+    return (PreparedStatement) PreparedStatement.of(
+        SqlFormatter.formatSql(statementWithSchema),
+        statementWithSchema);
+  }
+
+  /**
+   * Resolves all topic properties from the statement so that topic name, number partitions
+   * and number of replicas are explicitly specified.
+   *
+   * @return a prepared statement with the resolved properties and updated statement text
+   */
+  public static PreparedStatement withInferredSinkTopic(
+      final PreparedStatement statement,
+      final Map<String, Object> propertyOverrides,
+      final KsqlConfig ksqlConfig
+  ) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final CreateAsSelect cas = (CreateAsSelect) statement.getStatement();
+
+    final String topic = topicName(cas, ksqlConfig);
+    final int partitions = numPartitions(cas, ksqlConfig, propertyOverrides);
+    final short replicas = numReplicas(cas, ksqlConfig, propertyOverrides);
+
+    final Map<String, Expression> properties = new HashMap<>(cas.getProperties());
+    properties.put(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY, new StringLiteral(topic));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_REPLICAS, new IntegerLiteral(replicas));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_PARTITIONS, new IntegerLiteral(partitions));
+
+    final CreateAsSelect withTopic = cas.copyWith(properties);
+    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
```

Comment:
formatSql should include the semicolon

Reply:
Never mind!

Reply:
Actually no, this should already have the semicolon.

Reply:
For some reason, CSAS/CTAS does not:
```java
    final CreateStreamAsSelect csas = new CreateStreamAsSelect(
        QualifiedName.of("stream"),
        new Query(
            new QuerySpecification(
                new Select(true, ImmutableList.of(new AllColumns(new NodeLocation(0, 0)))),
                new Table(QualifiedName.of("sink")),
                true,
                new Table(QualifiedName.of("source")),
                Optional.empty(),
                Optional.empty(),
                Optional.empty(),
                Optional.empty(),
                OptionalInt.empty()
            ),
            OptionalInt.empty()
        ),
        false,
        ImmutableMap.of(),
        Optional.empty()
    );
    final String sql = SqlFormatter.formatSql(csas);
    System.out.println(sql); 
    // > CREATE STREAM stream AS SELECT DISTINCT *\nFROM TABLE source
```

Reply:
Seems like a bug to me. Don't want to block this patch on it though.